### DISCUSSION TITLE
base points associated to dof

### DISCUSSION THREAD

**Discussion by edljk** [Ferrite-FEM/Ferrite.jl#476] [Q&A] [answered]
Thank you for developing `Ferrite.jl`!
I am using Lagrange finite elements.  
In that case, is it possible for every dof to get the coordinates of a point representing the associated basis element. I am thinking to a function analogous to the following in [getfem++](https://getfem.org/scilab/cmdref_gf_mesh_fem_get.html)

```
DOFpts = gf_mesh_fem_get(mesh_fem MF, 'basic dof nodes'[, mat DOFids])
Get location of basic degrees of freedom.
```
Thanks

**Comment by fredrikekre** [comment]
If I understand you correctly you want to find the coordinate `x` for a given input dof `y`? I don't think there is an easy way to do that right now, but probably a hacky way. Can I ask what you need it for?

**Comment by edljk** [comment]
You are welcome! I am working with polygons / polytops with strange periodicity conditions. To enforce the global regularity of a fem function on such domains, I need to  be able to merge dofs which are equivalent up to these periodic conditions. I understand that this is a non standard use of fem..
Is the motivation clear enough?

**Comment by edljk** [comment]
Does the following hack (inspired by test functions) looks reasonable to you?

```julia
function dof_points(grid, qr, ip, ip_geom)
    dim = length(grid.nodes[1].x)
    fx = x -> x[1]
    fy = x -> x[2]
    valsx = _analytical(grid, qr, ip, ip_geom, fx)
    valsy = _analytical(grid, qr, ip, ip_geom, fy)
    L2p = L2Projector(ip, grid, geom_ip = ip_geom)
    xdof = project(L2p, valsx, qr, project_to_nodes = false)
    ydof = project(L2p, valsy, qr, project_to_nodes = false)
    if dim == 2
        return hcat(xdof, ydof)
    else
        fz = x -> x[3]
        valsz = _analytical(grid, qr, ip, ip_geom, fz)
        zdof = project(L2p, valsz, qr, project_to_nodes = false)
        return hcat(xdof, ydof, zdof)
    end
    nothing
end
function _analytical(grid, qr, ip, ip_geom, f)
    qp_values = []
    cv = CellScalarValues(qr, ip, ip_geom)
    for cell in CellIterator(grid)
        reinit!(cv, cell)
        r = [f(spatial_coordinate(cv, qp, getcoordinates(cell))) for qp in 
             1:getnquadpoints(cv)]
        push!(qp_values, r)
    end
    return identity.(qp_values)
end
```

**Comment by edljk** [comment]
Previous approach seems to be ok using P2 elements on linear (standard) triangular mesh

![fem2_mesh1](https://user-images.githubusercontent.com/14992507/189403236-15a70352-b822-4709-852b-dc7a95a48a88.png)

But for curvelinear mesh of order 2, even if it seems almost correct,
![fem2_mesh2](https://user-images.githubusercontent.com/14992507/189404254-85ff68fd-1c8f-4d11-9d2f-01745c5ee3b5.png)


 there is a small difference of coordinates between the initial vertices of triangles and the corresponding dof points. This should come from the L2 projection. Something should not be correct in my implementation, I expected  a perfect match at these points. See the zoom below

![fem2_mesh2_zoom](https://user-images.githubusercontent.com/14992507/189404418-2877cc37-d356-4aec-a52b-71f130963427.png)

**Comment by edljk** [comment]
I identified my mistake. I was using an inexact quadrature order: Switching from 3 to 4 for fem of degree 2 gives the good dof points also with  curved mesh (of order 2). Thanks.

**Comment by fredrikekre** [comment, marked as answer]
There is probably a more efficient way of doing it, but great that you got it working!

Note that there is some support for periodic boundary conditions, where you don't have to do this manipulation manually like this: https://ferrite-fem.github.io/Ferrite.jl/stable/manual/boundary_conditions/#Periodic-boundary-conditions but currently it only works for hypercubes. I am currently generalizing though, so should be much simpler in the near future!
